File: src/main/java/org/waterpouring/Solver.java
package org.waterpouring;

import java.util.*;

public class Solver {
    final Integer[] JUGS;
    final Integer GOAL;

    // The initial state we begin with (all jugs empty)
    final Integer[] START_STATE;

    /**
     *
     * @param jugs The capacity, in gallons, of the jugs we are working with
     * @param goal in gallons, of the desired amount
     */
    public Solver(final Integer[] jugs, final Integer goal) {
        this.JUGS = jugs;
        this.GOAL = goal;
        this.START_STATE = new Integer[JUGS.length];
        for(int i = 0; i< START_STATE.length; i++) {
            this.START_STATE[i] = 0;
        }
    }

    /**
     * For the given state, return all possible actions that can be taken
     * @param state
     * @return
     */
    public List<Action> createActions(Integer[] state) {
        List<Action> pourStates = new ArrayList<>();
        for(int i=0; i<JUGS.length; i++) {
            pourStates.add(new Action(state, MoveType.EMPTY, i));
            pourStates.add(new Action(state, MoveType.FILL, i));
            for(int j=0; j<JUGS.length; j++) {
                pourStates.add(new Action(state, MoveType.POUR, i, j));
            }
        }
        return pourStates;
    }

    public Integer[] renderAction(Action action) {
        if(action.getMoveType()==null) return START_STATE;
        return action.getMoveType().getMove()
                .doMove(JUGS, action.getState(), action.getJugIndex());
    }

    /**
     * Look within the tree of all possible actions to see if a goal solution exists
     * @param actions
     * @return
     */
    public boolean solutionExists(List<List<Action>> actions) {
        for(List<Action> trail : actions) {
            if(solutionExistsInTrail(trail)) {
                return true;
            }
        }
        return false;
    }

    private boolean solutionExistsInTrail(List<Action> actionTrail) {
        Action lastAction = actionTrail.get(actionTrail.size() - 1);
        Integer[] endState = renderAction(lastAction);
        for(int i=0; i < endState.length; i++) {
            if(endState[i] == GOAL) {
                return true;
            }
        }
        return false;
    }

    /**
     * For the given trail, take all the possible actions in the next list and create
     * a new list for each one where the trail of actions becomes the head of the list,
     * and the next element is appended to the end.
     * @param trail of actions
     * @param next possible next actions from the end of the trail.
     * @return For example: <p>
     *     trail = "A", "B", "C" and next = "D", "E", "F"
     * <p>yields:
     * <ul>
     *     <li> "A", "B", "C", "D"</li>
     *     <li> "A", "B", "C", "E"</li>
     *     <li> "A", "B", "C", "F"</li>
     * </ul>
     */
    public List<List<Action>> combineTrails(List<Action> trail, List<Action> next) {
        List<List<Action>> retList = new ArrayList<>();
        for(Action nextAction : next) {
            List<Action> updated = new ArrayList<>(trail);
            updated.add(nextAction);
            retList.add(updated);
        }
        return retList;
    }

    /**
     * Helper method that takes an application state and return new actions
     * that can be performed from it
     * @param action The current state of jugs (ie, how much water is in each one)
     * @param stateHistory The list of states we have already visited
     * @return The list of new actions that can be taken
     */
    private List<Action> solve(Action action, Set<List<Integer>> stateHistory) {
        Integer[] state = renderAction(action);
        List<Action> retState = createActions(state);
        List<Action> retActions = new ArrayList<>();
        for (Action nextAction : retState) {
            Integer[] nextState = renderAction(nextAction);
            if (stateHistory.contains(Arrays.asList(nextState))) continue;
            stateHistory.add(Arrays.asList(nextState));
            retActions.add(nextAction);
        }
        return retActions;
    }

    /**
     * For the given parameters, find a solution where the Jug capacities can produce
     * the goal amount.
     * @return A list of actions that yield the goal amount, or null if no solution exists
     */
    public List<Action> findSolution() {
        Set<List<Integer>> history = new HashSet<>();
        history.add(Arrays.asList(START_STATE));
        List<List<Action>> trails = new ArrayList<>();

        List<Action> initialTrail = new ArrayList<>();
        initialTrail.add(new Action(START_STATE, null, new Integer[]{}));
        trails.add(initialTrail);

        while(!solutionExists(trails) && trails.size() > 0) {
            ListIterator<List<Action>> it = trails.listIterator();
            while(it.hasNext()) {
                List<Action> trail = it.next();
                Action endState= trail.get(trail.size()-1);
                List<Action> next = solve(endState, history);
                it.remove();

                if(next.size() > 0) {
                    List<List<Action>> combined = new ArrayList<>();
                    combined.addAll(combineTrails(trail, next));
                    for(List<Action> combinedTrail : combined) {
                        it.add(combinedTrail);
                    }
                }
            }
        }

        for(List<Action> trail : trails) {
            if(solutionExistsInTrail(trail)) return trail;
        }
        return null;
    }

    public static void main(String[] args) {
        List<Action> solution = new Solver(new Integer[]{5, 3}, 4).findSolution();
        if(solution != null) {
            System.out.println("Solution found: " + solution);
        } else {
            System.out.println("No solution found for the given parameters");
        }
    }
}

File: src/main/scala/Solver.scala
object WaterPouringTypes {
  type Jug = Int
  type State = List[Int]
}
import WaterPouringTypes._

class WaterPouring(val JUGS: List[Jug] = List(3, 4), val GOAL: Integer = 2) {
  val INITIAL_STATE: State = JUGS.map(_=>0) //All jugs are empty to begin with
  val JUG_INDEX_RANGE = 0 to JUGS.length - 1 toList //Index iterator over all jugs

  trait Move {
    def doMove(state: State): State
  }
  case class Empty(jugIndex: Jug) extends Move {
    override def doMove(state: State): State = state updated (jugIndex, 0)
  }
  case class Fill(jugIndex: Jug) extends Move {
    override def doMove(state: State): State = state updated (jugIndex, JUGS(jugIndex))
  }
  case class Pour(fromJugIndex: Jug, toJugIndex: Jug) extends Move {
    override def doMove(state: State): State = {
      val pourAmt = math.min(state(fromJugIndex), JUGS(toJugIndex) - state(toJugIndex))
      state updated (fromJugIndex, state(fromJugIndex) - pourAmt) updated (toJugIndex, state(toJugIndex) + pourAmt)
    }
  }

  class Path(val moves: List[Move], val endState: State) {
    override def toString: String = moves + " yielding " + endState
  }

  val EMPTY_MOVES = for (jug <- JUG_INDEX_RANGE) yield Empty(jug)
  val FILL_MOVES = for(jug <- JUG_INDEX_RANGE) yield Fill(jug)
  val POUR_MOVES = for {
    fromJug <- JUG_INDEX_RANGE
    toJug <- JUG_INDEX_RANGE
    if fromJug != toJug
  } yield Pour(fromJug, toJug)
  val MOVES: List[Move] = EMPTY_MOVES ++ FILL_MOVES ++ POUR_MOVES

  def createPaths(path: Path): Stream[(Move, State)] =
    MOVES.toStream.map(m => (m, m.doMove(path.endState)))

  def solve(paths: Stream[Path], history: Set[State]): Stream[Path] =
    paths match {
      case Stream.Empty => Stream()
      case head #:: xs => {
        val next = createPaths(head)
          .filter(t => !(history contains(t._2)))
          .map(t => new Path(head.moves :+ t._1, t._2))
        head #:: solve(xs #::: next, history | next.map(t=>t.endState).toSet)
      }
    }

  def findSolution() =
    solve(Stream(new Path(List(), INITIAL_STATE)), Set(INITIAL_STATE))
      .find(p => p.endState.contains(GOAL))
}

object Solver extends App {
  var w = new WaterPouring(List(3, 5), 4)
  val solution = w.findSolution()
  if(solution.isDefined)
    println("Yipeekayay! Found a solution at " + solution.get)
  else println("No solution found, John")

}
